### ISSUE TITLE
False Positive on Dataclass Transforms

### ISSUE THREAD

**Issue by Mercerenies** [microsoft/pyright#5674]
**Describe the bug**

When a dataclass (or dataclass transform) field lacks a type annotation, Pyright reports an error. That is, if I write

```python
from dataclasses import dataclass, field

# Pyright error, and fails at runtime.
@dataclass
class Example:
    x = field() # Look, Ma, no annotations!
```

Then Pyright correctly reports an error, as that field should have a type annotation. However, if I do the same thing in `attrs`,

```python
import attr

# Works at runtime, but Pyright still complains.
@attr.s
class Example:
    x = attr.ib() # Look, Ma, no annotations!
```

This is perfectly working code. No runtime errors. Yet Pyright still tells me "Dataclass field without type annotation will cause runtime exception", which is simply not true. I would expect the `attr.s` version of `Example` to pass type-checking and for `x` to get an implicit type of `Any`, to allow gradual introduction of Pyright into a codebase that uses `attr.s` without annotations today.

**Code or Screenshots**

(see above)

**VS Code extension or command-line**

Pyright 1.1.296 via command line

**Comment by erictraut** [comment]
Pyright's behavior here is correct. As you show, when you use a `dataclass`, you need to provide a type annotation for each field. The `dataclass_transform` applies the same rules. It has no special knowledge of `attrs` or its idiosyncratic behaviors — places where it deviates from `dataclass`.

If you want to use `attrs` with pyright, you'll need to stick to standard `dataclass` behaviors. In this case, that means including a type annotation for each field.

**Comment by bhearsum** [comment]
This may be by design, but the error message is simply incorrect for the `attr.s` example.
